src/main/scala/org/apache/spark/sql/kinesis/KinesisSourceProvider.scala

```diff
@@ -112,6 +113,20 @@ private[kinesis] class KinesisSourceProvider extends DataSourceRegister
       throw new IllegalArgumentException(
         "Stream name is a required field")
     }
+    if(caseInsensitiveParams.contains(STARTING_POSITION_KEY)) {
```

Comment:
Thanks for picking this up. We should have the support of a custom starting position. 
But shouldn't we have a JSON here with a starting sequence number for each shard?

Reply:
This is done to isolate from the existing logic to provide custom support for each shards.

Reply:
Can you give an example of how it will be used?  From what I understood, the starting sequence number would be same for all shards. But that is not correct. If we have to support it, we should make it flexible enough that individual shards can provide their own offset

Also, what is the strategy for new shards?

Reply:
The starting number is not the same for all the shards. The starting sequence number of the records to be read from each shard can be mentioned in the below format:

```.option("startingposition", "AFTER_SEQUENCE_NUMBER")
 .option("sequencenumber", """{"<shard_id-01>":"<sequence_number to be read from>"},{"<shard_id-02>":"<sequence_number>"}""" )```

Regarding the new shards, as mentioned above, the below strategy will be applicable:
**For the non-mentioned shards in the value of "sequencenumber":
a) Records will be processed from TRIM_HORIZON for open/new shards
b) Records won't be processed for the closed shards**

Reply:
This is a good enhancement. We should use same key `startingposition` to provide a custom positions. Just like we have for kafka - https://spark.apache.org/docs/latest/structured-streaming-kafka-integration.html

Also, we should break this PR into two - to support per-shard read positions position and the other one to better handle retryOrTimeout.

For per-shard read positions, can we include few unit tests as well?


--